File: app/src/main/java/com/xjm/xxd/dribbird/login/LoginActivity.java
package com.xjm.xxd.dribbird.login;

import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.StringRes;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.webkit.WebSettings;

import com.xjm.xxd.dribbird.R;
import com.xjm.xxd.dribbird.main.MainActivity;
import com.xjm.xxd.dribbird.base.BaseActivity;
import com.xjm.xxd.dribbird.base.BaseDialog;
import com.xjm.xxd.dribbird.widget.ProgressWebView;
import com.xjm.xxd.dribbird.utils.ToastUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class LoginActivity extends BaseActivity implements
        LoginActivityView {

    @BindView(R.id.tool_bar)
    Toolbar mToolbar;
    @BindView(R.id.web_view)
    ProgressWebView mWebView;

    ILoginActivityPresenter mPresenter;

    private BaseDialog mBaseDialog;
    private LoginWebViewClient mWebViewClient;

    private static final String TAG = LoginActivity.class.getSimpleName();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        ButterKnife.bind(this);
        mPresenter = new LoginActivityPresenter();

        initViews();

        mPresenter.bindIView(this);
        mPresenter.init();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mPresenter != null) {
            mPresenter.onDestroy();
        }

        if (mWebViewClient != null) {
            mWebViewClient.onDestroy();
        }
    }

    private void initViews() {
        mToolbar.setTitle(R.string.title_login);
        setSupportActionBar(mToolbar);

        WebSettings settings = mWebView.getSettings();
        settings.setJavaScriptEnabled(true);
        mWebViewClient = new LoginWebViewClient(mPresenter);
        mWebView.setWebViewClient(mWebViewClient);
    }

    @Override
    public void loadUrl(String url) {
        if (TextUtils.isEmpty(url)) {
            return;
        }
        mWebView.loadUrl(url);
    }

    @Override
    public void showLoading(String msg) {
        if (TextUtils.isEmpty(msg)) {
            return;
        }
        if (mBaseDialog == null) {
            BaseDialog.Builder builder = new BaseDialog.Builder();
            builder.progressBar(true);
            mBaseDialog = builder.build();
        }
        mBaseDialog.message(msg);
        mBaseDialog.show(getFragmentManager(), TAG);
    }

    @Override
    public void showLoading(@StringRes int strId) {
        if (strId <= 0) {
            return;
        }
        showLoading(getString(strId));
    }

    @Override
    public void hideLoading() {
        if (mBaseDialog != null && mBaseDialog.isVisible()) {
            mBaseDialog.dismiss();
        }
    }

    @Override
    public void loginSuccess() {
        ToastUtils.show(this, R.string.authentication_success);
        MainActivity.open(this);
        finish();
    }

    @Override
    public void loginFailed() {
        ToastUtils.show(this, R.string.authentication_failed);
    }

    public static void open(BaseActivity activity) {
        if (activity == null) {
            return;
        }
        Intent intent = new Intent(activity, LoginActivity.class);
        activity.startActivity(intent);
    }

}

File: app/src/main/java/com/xjm/xxd/dribbird/login/LoginActivity.kt
package com.xjm.xxd.dribbird.login

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.widget.Toolbar
import android.text.TextUtils
import com.xjm.xxd.dribbird.R
import com.xjm.xxd.dribbird.account.TokenManager
import com.xjm.xxd.dribbird.homepage.MainActivity
import com.xjm.xxd.dribbird.widget.ProgressWebView
import com.xjm.xxd.framework.kotlinext.bindView
import com.xjm.xxd.framework.widget.DialogFactory
import com.xjm.xxd.skeleton.kotlinext.toast
import com.xjm.xxd.skeleton.mvp.MVPActivity

class LoginActivity : MVPActivity<LoginActivityContract.Presenter, LoginActivityContract.Viewer>(),
        LoginActivityContract.Viewer {

    private val mToolbar by bindView<Toolbar>(R.id.tool_bar)
    private val mWebView by bindView<ProgressWebView>(R.id.web_view)

    private var mLoadingDialog: AlertDialog? = null
    private val mDialogFactory by lazy { DialogFactory() }

    private var mWebViewClient: LoginWebViewClient? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)

        initViews()

        loadUrl(TokenManager.oAuth2Url)
    }

    override fun createPresenter(): LoginActivityContract.Presenter = LoginActivityPresenter()

    @SuppressLint("SetJavaScriptEnabled")
    private fun initViews() {
        mToolbar.setTitle(R.string.title_login_activity)
        setSupportActionBar(mToolbar)

        val settings = mWebView.settings
        settings.javaScriptEnabled = true

        mWebViewClient = LoginWebViewClient(presenter())
        mWebView.webViewClient = mWebViewClient
    }

    override fun loadUrl(url: String?) {
        url?.let {
            mWebView.loadUrl(it)
        }
    }

    override fun showLoading(msg: String?) {
        if (TextUtils.isEmpty(msg)) {
            return
        }
        showLoadingDialog(msg)
    }

    private fun showLoadingDialog(msg: String?) {
        mLoadingDialog?.dismiss()
        mLoadingDialog = mDialogFactory.createLoadingDialog(this@LoginActivity, msg)
        mLoadingDialog?.show()
    }

    override fun hideLoading() {
        mLoadingDialog?.dismiss()
    }

    override fun loginSuccess() {
        toast(R.string.authentication_success)
        MainActivity.open(this)
        finish()
    }

    override fun loginFailed() {
        toast(R.string.authentication_failed)
    }

    override fun onDestroy() {
        mLoadingDialog?.dismiss()
        mWebViewClient?.onDestroy()
        super.onDestroy()
    }

    companion object {

        const val LOADING_DIALOG_FRAGMENT_TAG = "LoadingDialog"

        @JvmStatic
        fun open(context: Context?) {
            context?.let {
                val intent = Intent(it, LoginActivity::class.java)
                it.startActivity(intent)
            }
        }
    }

}
